### ISSUE TITLE
[BUG] python3.10 spec.get_string_attribute() throws a runtime exception with non-utf-8 string headers

### ISSUE THREAD

**Issue by voodoonofx** [AcademySoftwareFoundation/OpenImageIO#3856]
**Describe the bug**

When looping over all key/value pairs in an EXR header using the python bindings to OpenImageIO, a field containing 0xFF values causes an awkward `RuntimeError: Could not allocate string object!`. This is very commonly seen in DI when the upstream writing tool doesn't want to set a value and instead just fills it with some default of `0xFF`.

**To Reproduce**
With the following header (as shown with iinfo -a -v):
```sh
    AudioFramerate: 24
    AudioInfo: "������������������������������������������������������������������������@"
    AudioRunningLTC: "MOS"
```
Where the `�` character is `0xFF`.

Attempting to iterate over headers like so:

```python
>>> import OpenImageIO  as oiio
>>> src = oiio.ImageBuf('./my_file.1000.exr')
>>> spec = src.spec()
>>> for a in spec.extra_attribs:
...   print(f'{a.name} ({a.type}): {a.value} ({type(a.value)})')
```

Causes the following exception:

```python
AudioFileName (string):  (<class 'str'>)
AudioFramerate (float): 24.0 (<class 'float'>)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
RuntimeError: Could not allocate string object!
```

**Expected behavior**
It appears that strings are assumed to be in utf-8 unicode, which causes an error trying to decode non-unicode data into a python string.

The workaround is to not access `a.value` and instead ask for the attribute directly like so:

```python
>>> try:
...   x = spec.get_string_attribute('AudioInfo')
... except RuntimeError:
...   pass
```

However, I was hoping to have a more graceful exception of this error. For example, I cannot access the raw bytes of this value to determine if the offending field is actually usable. Perhaps adding an `a.raw_value` could return the information in the field as a bytestring so the above try/except has an opportunity to do something else special with the value. Since the accessing of an attribute here is forcing the coercion to unicode, I cannot get to the original data using python in any way.

**Platform information:**
 - OIIO branch/version:  `openimageio-2.4.5.0-r0`
 - OS:  `Alpine Linux 3.17 under Docker`
 - C++ compiler:  
 - Any non-default build flags when you build OIIO:

**Comment by voodoonofx** [comment]
To work around this, I just switched to using `spec.get_bytes_attribute()`, and then attempting my own string conversions.

However, the python `get_string_attribute` should still not fail egregiously even if the underlying value cannot be parsed as a string. Perhaps just raising the `UnicodeDecodeError` directly would allow a try/except to catch this rather than the much more generic `RuntimeError`

**Comment by lgritz** [comment]
This could be a good first issue for somebody who is fluent in these python expectations and knows how to fix this.

**Comment by nrusch** [comment]
I'll raise my hand to take a look at this (assuming I can figure out how to reproduce it...)

**Comment by nrusch** [comment]
OK, I've got this narrowed down, so I guess the remaining question is how best to handle it.

Given that the Python API is more or less entirely duck-typed with respect to attributes, I think the "simple" answer might be to try and perform the default `py:str` cast (with its implicit UTF-8 decoding), and if that fails, we just return a Python `bytes` object.

I'll put together a PR with some tests, but if anyone else has any other ideas or opinions, please feel free to share.

**Comment by lgritz** [comment]
If this is a call specifically to get_string_attribute, I think it should always return a valid str.

Let's back up a little. I'm trying to understand what's happening upstream.

There is an exr attribute being saved. What is the data type of it in the exr file? What data type does our exr reader use when storing it at an OIIO attribute in the ParamList? 

Does EXR have an opinion about whether string attributes are UTF-8?

**Comment by nrusch** [comment]
> If this is a call specifically to get_string_attribute, I think it should always return a valid str.

It's not only that call. In addition to `get_string_attribute`, this applies to `ImageSpec.channel_name`, as well as any value access on a `ParamValue` holding a `std::string` or `const char*`. It also applies to `ImageSpec.to_xml`, but I probably wouldn't try to lump that into this effort.

Currently, all of the Python value conversions for `std::string` and `const char*` are based on trying to instantiate `py::str` with no checks. Per the [pybind11 docs](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html#returning-c-strings-to-python), this will raise a decoding exception if the data is not valid UTF-8:

> When a C++ function returns a std::string or char* to a Python caller, pybind11 will assume that the string is valid UTF-8 and will decode it to a native Python str, using the same API as Python uses to perform bytes.decode('utf-8'). If this implicit conversion fails, pybind11 will raise a UnicodeDecodeError.

---

> There is an exr attribute being saved. What is the data type of it in the exr file? What data type does our exr reader use when storing it at an OIIO attribute in the ParamList?

> Does EXR have an opinion about whether string attributes are UTF-8?

The data is a `string` attribute in the EXR file, and likewise when read in OIIO. The OpenEXR code states that filenames and some (all?) standard attribute values should be/are interpreted as being UTF-8 encoded, but I can't find anything that says the same about arbitrary user attributes.

In other words, I think all of this is totally above board from the perspective of both libraries, and no one is doing anything "wrong" by creating a file with a non-UTF-8 string attribute. So even if you wanted `get_string_attribute` to always return a valid `str` (the actual Python type), there is no singular definition of "valid `str`" to apply here.

P.S. Just in case it's useful, creating an attribute like this is trivial using the OIIO bindings:
```python
spec.attribute("bytestr", b"\xFF")
```

**Comment by nrusch** [comment]
> So even if you wanted `get_string_attribute` to always return a valid `str` (the actual Python type), there is no singular definition of "valid `str`" to apply here.

Just to elaborate on this, there are a number of different options for how to handle non-UTF-8 data:
1. Return a `bytes` object (my original thought), so at least the user has a pure representation of the underlying data.
2. Perform the decoding using one of the error handling strategies listed at https://docs.python.org/3.11/library/codecs.html#error-handlers (besides `'strict'`).

To give you an idea of what these alternate error handling strategies would do, here's what you get with each one for the 2-byte sequence `0xff 0xff` (`b"\xff\xff"` in Python):
```
replace: '��'
ignore: ''
backslashreplace: '\\xff\\xff'
surrogateescape: '\udcff\udcff'
```

**Comment by lgritz** [comment]
I feel strongly that get_attribute_string, or other things clearly advertised or implied by name as returning strings, ought to return a str, and not "maybe a str or maybe a bytes".

But beyond that, the best way to detect valid strings or convert a C++ string into a Python str, is not something I feel like I know enough about Python conventions, and am open to your best advice.

I should just throw this wrench, though: Now that we're switching from pybind11 to nanobind, does that change the situation at all?

**Comment by nrusch** [comment]
> I feel strongly that get_attribute_string, or other things clearly advertised or implied by name as returning strings, ought to return a str, and not "maybe a str or maybe a bytes".

Fair enough. I think the best option here might be the `'surrogateescape'` strategy, as it can be used symmetrically to re-encode the `str` to `bytes` if needed. I can put together a PR if you think that sounds acceptable.

It may be nice to look at adding a direct bytes accessor to the Python API as well, though that's something that could be done separately.

> I should just throw this wrench, though: Now that we're switching from pybind11 to nanobind, does that change the situation at all?

I don't think so. From a quick scan of the source, it looks like `nanobind` effectively uses the same decoding logic as `pybind11`.

**Comment by lgritz** [comment]
Yes, I would appreciate if you could distill the best practice here into a PR. It's outside my expertise, so I trust your judgment here. My bottom line is:

* OIIO API calls that semantically return a string (implied by their name, functionality, or symmetry with the C++ APIs) should return a real Python `str` every time.
* Python scripts assuming that a `str` returned is valid should not have to catch exceptions, check validity, or re-encode just in case for every string we give them.
* So in our python bindings themselves that pass back to Python "uncontrolled" strings (such as things read from files), we should do... whatever... makes them safe for Python. There may be plenty of our API calls that pass strings but that we can be certain will be safe, and we don't need to re-scan or re-encode those.

**Comment by nrusch** [comment]
I cooked up a solution against dev-3.1 that works well, but in looking at adapting it to `main`, it looks like the previous template struct pattern has been replaced with plain `py::cast` everywhere (with the exception of `half` types).

Assuming it's worth backporting this fix to the 3.1 line, would you prefer I start by targeting `main` as normal and then we can handle the backporting separately? I think that patch will look substantially different, so it may need its own PR.

**Comment by lgritz** [comment]
Yeah, main (and 3.2, which hasn't yet truly divereged) has had some refactoring of the python bindings in order to support the switch to nanobind, compared to 3.1.

We are literally days from 3.2 becoming the supported release, and 3.1 being considered obsolete -- getting continued patches for a while, but becoming increasingly stingy over time about what we bother to backport.

Since this string validity issue has gone unnoticed until now, I would support just implementing it in main/3.2 and pushing off a 3.1 solution until somebody needs it (and even then, have a hard look at whether it's worth backporting or just telling people they should use 3.2 if they want a fix).

**Comment by nrusch** [comment]
OK, I've got a draft PR up here with a working approach: https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/5485

I still need to add some tests, but I at least wanted to get your thoughts on the general approach.

I've left `ImageSpec.to_xml` and `ImageSpec.serialize` alone for now, but I think they will need to be addressed, as they are currently prone to the same failure condition. However, I want to better understand the implications for XML parsing specifically, because a simple test of wrapping `to_xml` produced a result that couldn't be parsed by Python's `etree` API.
